### PR TITLE
Auto style comm

### PR DESCRIPTION
**Pull request by donflopez**
This PR add:

- Custom colors for widgets
- Custom auto-style feature
- Enable/Disable autostyle button
- Edit/delete button to dropdown
- Dynamic color changes and autostyle scheme
- Fix when cartocss has conditional layer
- Dark custom colors
- Some basic styling features unified under widget-model (it need more work here)
- Method to return definitions with ramp and domains

### CODE REVIEW

**Comment by donflopez** [comment]
This PR is part of [CartoDB/cartodb#10255](https://github.com/CartoDB/cartodb/issues/10255)

File: package.json [hunk 1]
```diff
@@ -23,11 +23,13 @@
   "license": "BSD-3-Clause",
   "dependencies": {
     "backbone": "1.2.3",
+    "cartocolor": "^3.0.0",
```
Comment (nobuti) [inline package.json]: I feel we should make this dependency fixed. Let it open could bring some troubles in the future.

File: package.json [hunk 2]
```diff
@@ -23,11 +23,13 @@
...
     "cartodb.js": "CartoDB/cartodb.js#v4",
     "d3": "3.5.17",
     "jquery": "2.1.4",
     "moment": "2.10.6",
     "perfect-scrollbar": "0.6.7",
+    "tinycolor2": "^1.4.1",
```
Comment (nobuti) [inline package.json]: Same as above.

File: src/widgets/auto-style/category-colors.js [hunk 3]
```diff
@@ -15,12 +25,17 @@ var colorScales = [[
  *
  */
 
-function CategoryColors () {
+function CategoryColors (styles) {
+  this.updateColors(styles);
+}
+
+CategoryColors.prototype.updateColors = function updateColors (styles) {
```
Comment (nobuti) [inline src/widgets/auto-style/category-colors.js]: do we need a named function? or `CategoryColors.prototype.updateColors = function (styles) {` is enough?

File: src/widgets/auto-style/category-colors.js [hunk 4]
```diff
@@ -51,6 +66,7 @@ CategoryColors.prototype.getNextAvailableColor = function () {
 };
 
 CategoryColors.prototype.getColorByCategory = function (category) {
+  // console.trace('Get color by cat');
```
Comment (nobuti) [inline src/widgets/auto-style/category-colors.js]: Delete it if not needed

File: src/widgets/auto-style/category.js [hunk 5]
```diff
@@ -14,23 +14,61 @@ module.exports = AutoStyler.extend({
     return StyleUtils.replaceWrongSpaceChar(style);
   },
 
+  getRange: function () {
+    var model = this.dataviewModel;
+    var categories = model.get('data');
+    var range = [];
```
Comment (nobuti) [inline src/widgets/auto-style/category.js]: Not wrong, just a matter of taste but why not writing this in a more functional style:
```
return _.map(categories, function (category) {
    return this.colors.getColorByCategory(category.name);
}, this)
```

File: src/widgets/auto-style/style-utils.js [hunk 6]
```diff
@@ -47,6 +59,9 @@ function changeStyle (cartocss, attr, newStyle) {
 }
 
 module.exports = {
-  changeStyle: changeStyle,
+  changeStyle: _.memoize(changeStyle, function (css, attr, style) {
```
Comment (nobuti) [inline src/widgets/auto-style/style-utils.js]: ❤️

File: package.json (hunk 1 above)
Comment (donflopez) [inline package.json]: Sure!! Thanks

File: src/widgets/widget-model.js [hunk 7]
```diff
@@ -18,8 +19,30 @@ module.exports = cdb.core.Model.extend({
     'collapsed': false
   },
 
+  super: function (name) {
```
Comment (nobuti) [inline src/widgets/widget-model.js]: Why do we need this? It looks like a serious hack.

File: src/widgets/widget-model.js [hunk 8]
```diff
@@ -44,6 +67,83 @@ module.exports = cdb.core.Model.extend({
...
+    return true;
+  },
+
+  reapplyAutoStyle: function () {
+    var style = this.autoStyler.getStyle();
+    this.dataviewModel.layer.set('cartocss', style);
+    this.set('autoStyle', true);
+  },
+
+  cancelAutoStyle: function (noRestore) {
```
Comment (nobuti) [inline src/widgets/widget-model.js]: As @alonsogarciapablo told me this morning, from a reader point of view, it's better passing an options objects with keys:
```
cancelAutoStyle: function (opts) {
   if (opts.noRestore === false) {
   ...
   }
}
```

File: src/widgets/widget-model.js (hunk 7 above)
Comment (donflopez) [inline src/widgets/widget-model.js]: hahaha It is! As I see the inheritance pattern, I've tried to implement the super methos (it is a super hack ;P) but node 0.10.XX can't run it :( so the tests are broken and I don't use it... so it is unnecessary. Let's remove it!

**Comment by nobuti** [comment]
It's big but it looks promising 👍

File: src/widgets/widget-model.js (hunk 8 above)
Comment (donflopez) [inline src/widgets/widget-model.js]: Why? I don't understand, sorry

Comment (alonsogarciapablo) [inline src/widgets/widget-model.js]: If we only have one flag now, we could also have:

```
cancelAutoStyleAndRestoreCartoCSS: function () {
  this.dataviewModel.layer.restoreCartoCSS();
  this.cancelAutoStyle();
},
cancelAutoStyle: function () {
  this.set('autoStyle', false);
}
```

Comment (donflopez) [inline src/widgets/widget-model.js]: Yeah! I see now! It's a way to force the caller to write the name of the variable and make more readable. I was seeing only the function side, not the caller one.

File: src/widgets/widget-model.js [hunk 9]
```diff
@@ -44,6 +55,83 @@ module.exports = cdb.core.Model.extend({
...
+      if (!initialStyle && layer.get('meta')) {
+        initialStyle = layer.get('meta').cartocss;
+      }
+      layer.set('initialStyle', initialStyle);
+    }
+
+    var style = this.autoStyler.getStyle();
+    layer.set('cartocss', style);
+    this.set('autoStyle', true);
+    // this.getAutoStyle();
```
Comment (nobuti) [inline src/widgets/widget-model.js]: Delete it if we don't need it anymore

**Comment by nobuti** [comment]
🇪🇸

File: src/widgets/widget-model.js (hunk 9 above)
Comment (donflopez) [inline src/widgets/widget-model.js]: Yeah! Sorry :)

**Pull request merged**